### PR TITLE
[NIS-1412] Impliment Office

### PR DESCRIPTION
**Pull request by NikDub**
[NIS-1412](https://innowise-group.atlassian.net/browse/NIS-1412)

### CODE REVIEW

File: OfficeMicroService/Application/Extensions/ServiceExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,84 @@
+﻿using Microsoft.AspNetCore.Authentication.JwtBearer;
+using Microsoft.Extensions.Options;
+using Microsoft.IdentityModel.Tokens;
+using Microsoft.OpenApi.Models;
+using OfficeMicroService.Application.Middlewares;
+using OfficeMicroService.Application.Services;
+using OfficeMicroService.Data.Mapper;
+using OfficeMicroService.Data.Settings;
+using Serilog;
+using Serilog.Events;
```
Comment (kseniya-kalita) [inline OfficeMicroService/Application/Extensions/ServiceExtensions.cs]: remove unnecessary usings

File: OfficeMicroService/Application/Extensions/ServiceExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+{
+    public static class ServiceExtensions
+    {
+        public static void ConfigureJWTAuthentication(this IServiceCollection services, IConfiguration configuration)
+        {
+            services.AddAuthentication(JwtBearerDefaults.AuthenticationScheme)
+                   .AddJwtBearer(JwtBearerDefaults.AuthenticationScheme, options =>
+                   {
+                       options.Authority = configuration.GetValue<string>("Routes:AuthorityRoute");
+                       options.Audience = "TestsAPI";
```
Comment (kseniya-kalita) [inline OfficeMicroService/Application/Extensions/ServiceExtensions.cs]: Find better place for this hardcoded string

File: OfficeMicroService/Application/Extensions/ServiceExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+                       options.TokenValidationParameters = new TokenValidationParameters
+                       {
+                           ValidateAudience = false
```
Comment (kseniya-kalita) [inline OfficeMicroService/Application/Extensions/ServiceExtensions.cs]: Think about security..

File: OfficeMicroService/Application/Middlewares/ExceptionHandlingMiddleware.cs [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        {
+            try
+            {
+                await _next(httpContext);
+            }
+            catch (Exception ex)
+            {
+                await HandleExceptionAsync(httpContext, ex);
+            }
+        }
```
Comment (kseniya-kalita) [inline OfficeMicroService/Application/Middlewares/ExceptionHandlingMiddleware.cs:25]: empty lines after & before

File: OfficeMicroService/Application/Middlewares/ExceptionHandlingMiddleware.cs [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+        private async Task HandleExceptionAsync(HttpContext context, Exception exception)
+        {
+            context.Response.ContentType = "application/json";
+            context.Response.StatusCode = exception switch
+            {
+                NotFoundException => StatusCodes.Status404NotFound,
+                UnauthorizedException => StatusCodes.Status401Unauthorized,
+                _ => StatusCodes.Status500InternalServerError
+            };
+            await context.Response.WriteAsync(new ErrorDetails()
```
Comment (kseniya-kalita) [inline OfficeMicroService/Application/Middlewares/ExceptionHandlingMiddleware.cs:38]: Should we log the Exception here?

File: OfficeMicroService/Application/Services/IOfficeServices.cs [hunk 6]
```diff
@@ -0,0 +1,18 @@
+﻿using OfficeMicroService.Data.Enum;
+using OfficeMicroService.Data.Models.DTO;
+
+namespace OfficeMicroService.Application.Services
+{
+    public interface IOfficeServices
+    {
+        Task<List<OfficeIdDTO>> GetAsync();
+        Task<OfficeIdDTO> GetAsync(string id);
```
Comment (kseniya-kalita) [inline OfficeMicroService/Application/Services/IOfficeServices.cs]: rename this method or the method above to better suit the context (the problem that naming is the same for one entity and list)

File: OfficeMicroService/Application/Services/OfficeServices.cs [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+using MongoDB.Driver;
+using OfficeMicroService.Data.Enum;
+using OfficeMicroService.Data.Models.DTO;
+using OfficeMicroService.Data.Settings;
+
+namespace OfficeMicroService.Application.Services
+{
+    public class OfficeServices : IOfficeServices
+    {
+        private readonly IMongoCollection<OfficeIdDTO> _offices;
```
Comment (kseniya-kalita) [inline OfficeMicroService/Application/Services/OfficeServices.cs]: Entities are used for databases, not DTOs

File: OfficeMicroService/Application/Services/OfficeServices.cs [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+        public async Task<List<OfficeIdDTO>> GetAsync()
+        {
+            return (await _offices.FindAsync(office => true)).ToList();
+        }
+
+        public async Task<OfficeIdDTO> GetAsync(string id)
+        {
+            try
+            {
+                return (await _offices.FindAsync(office => office.Id == id)).FirstOrDefault();
```
Comment (kseniya-kalita) [inline OfficeMicroService/Application/Services/OfficeServices.cs]: When you'll have exceptions? Can you avoid try-catch here?

File: OfficeMicroService/Application/Services/OfficeServices.cs [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+            }
+            catch(Exception)
+            {
+                return null;
+            }
+        }
+
+        public async Task<OfficeIdDTO> CreateAsync(OfficeDTO model)
+        {
+            var mapModel = _mapper.Map<OfficeIdDTO>(model);
```
Comment (kseniya-kalita) [inline OfficeMicroService/Application/Services/OfficeServices.cs]: service layer should also contain null-checks

File: OfficeMicroService/Data/Models/DTO/OfficeDTO.cs [hunk 10]
```diff
@@ -0,0 +1,13 @@
+﻿namespace OfficeMicroService.Data.Models.DTO
+{
+    public class OfficeDTO
+    {
+        public string Photo_id { get; set; }
```
Comment (kseniya-kalita) [inline OfficeMicroService/Data/Models/DTO/OfficeDTO.cs]: not c# field-style

File: OfficeMicroService/Data/Models/DTO/OfficeDTO.cs [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+        public string Status { get; set; }
```
Comment (kseniya-kalita) [inline OfficeMicroService/Data/Models/DTO/OfficeDTO.cs]: change type to adjust confluence

File: OfficeMicroService/Data/Models/DTO/OfficeDTO.cs [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+        public string City { get; set; }
```
Comment (kseniya-kalita) [inline OfficeMicroService/Data/Models/DTO/OfficeDTO.cs]: please add required attribute where necessary

File: OfficeMicroService/Data/Models/DTO/OfficeDTO.cs [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+        public string Street { get; set; }
+        public string HouseNumber { get; set; }
+        public string OfficeNumber { get; set; }
+        public string RegistryPhoneNumber { get; set; }
```
Comment (kseniya-kalita) [inline OfficeMicroService/Data/Models/DTO/OfficeDTO.cs]: not enough checks

File: OfficeMicroService/Data/Models/DTO/OfficeIdDTO.cs [hunk 14]
```diff
@@ -0,0 +1,13 @@
+﻿using MongoDB.Bson;
+using MongoDB.Bson.Serialization.Attributes;
+
+namespace OfficeMicroService.Data.Models.DTO
+{
+    public class OfficeIdDTO : OfficeDTO
```
Comment (kseniya-kalita) [inline OfficeMicroService/Data/Models/DTO/OfficeIdDTO.cs]: find better name

File: OfficeMicroService/Presentation/Controllers/OfficeController.cs [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+using OfficeMicroService.Application.Services;
+using OfficeMicroService.Data.Enum;
+using OfficeMicroService.Data.Models.DTO;
+using Serilog;
+
+namespace OfficeMicroService.Presentation.Controllers
+{
+    [Route("[controller]/[action]")]
+    [ApiController]
+    public class OfficeController : Controller
```
Comment (kseniya-kalita) [inline OfficeMicroService/Presentation/Controllers/OfficeController.cs]: plural

File: OfficeMicroService/OfficeMicroService.csproj [hunk 16]
```diff
@@ -0,0 +1,33 @@
+<Project Sdk="Microsoft.NET.Sdk.Web">
+
+  <PropertyGroup>
+    <TargetFramework>net7.0</TargetFramework>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <GenerateDocumentationFile>True</GenerateDocumentationFile>
+    <DocumentationFile>C:\Users\nikdu\Desktop\git\OfficeMicroService\OfficeMicroService\Comments.xml</DocumentationFile>
```
Comment (kseniya-kalita) [inline OfficeMicroService/OfficeMicroService.csproj]: try to make it relative so everyone could run

File: OfficeMicroService/Presentation/Controllers/OfficeController.cs [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Mvc;
+using OfficeMicroService.Application.Services;
+using OfficeMicroService.Data.Enum;
+using OfficeMicroService.Data.Models.DTO;
+using Serilog;
+
+namespace OfficeMicroService.Presentation.Controllers
+{
+    [Route("[controller]/[action]")]
```
Comment (kseniya-kalita) [inline OfficeMicroService/Presentation/Controllers/OfficeController.cs]: having verbs in endpoints name violates rest

File: OfficeMicroService/Presentation/Controllers/OfficeController.cs [hunk 18]
```diff
@@ -0,0 +1,129 @@
...
+            _officeServices = officeServices;
+        }
+
+        /// <summary>
+        /// Get All Offices
+        /// </summary>
+        /// <response code="200">Returns offices list</response>
+        /// <response code="500">Operation wasn't succeeded</response>
+        [HttpGet]
+        public async Task<IActionResult> GetAll()
```
Comment (kseniya-kalita) [inline OfficeMicroService/Presentation/Controllers/OfficeController.cs:28]: create a separate dto to fit the confluence

File: OfficeMicroService/Presentation/Controllers/OfficeController.cs [hunk 19]
```diff
@@ -0,0 +1,129 @@
...
+        {
+            Log.Information("Method GetAll");
```
Comment (kseniya-kalita) [inline OfficeMicroService/Presentation/Controllers/OfficeController.cs]: find better approach then just typing method's name

File: OfficeMicroService/Presentation/Controllers/OfficeController.cs [hunk 20]
```diff
@@ -0,0 +1,129 @@
...
+        /// Create office for Receptionist
+        /// </summary>
+        /// <param name="model"></param>
+        /// <response code="200">Returns created office</response>
+        /// <response code="401">Unauthorized</response>
+        /// <response code="403">Forbidden</response>
+        /// <response code="500">Operation wasn't succeeded</response>
+        [Authorize(Roles = nameof(UserRole.Receptionist))]
+        [HttpPost]
+        public async Task<IActionResult> Create(OfficeDTO model)
```
Comment (kseniya-kalita) [inline OfficeMicroService/Presentation/Controllers/OfficeController.cs]: where will the logic of creating a photo be?

File: OfficeMicroService/Presentation/Controllers/OfficeController.cs [hunk 21]
```diff
@@ -0,0 +1,129 @@
...
+        {
+            Log.Information("Method Create {0} {1} {2} {3} {4}", model.City, model.Street, model.HouseNumber, model.RegistryPhoneNumber, model.OfficeNumber);
```
Comment (kseniya-kalita) [inline OfficeMicroService/Presentation/Controllers/OfficeController.cs]: probably it's better to parse this object to have the whole object with fields logged

File: OfficeMicroService/Presentation/Controllers/OfficeController.cs [hunk 22]
```diff
@@ -0,0 +1,129 @@
...
+            var office = await _officeServices.CreateAsync(model);
+            return Ok(office);
```
Comment (kseniya-kalita) [inline OfficeMicroService/Presentation/Controllers/OfficeController.cs]: find better status code

File: OfficeMicroService/Presentation/Controllers/OfficeController.cs [hunk 23]
```diff
@@ -0,0 +1,129 @@
...
+        }
+
+        /// <summary>
+        /// Update office for Receptionist
+        /// </summary>
+        /// <param name="id"></param>
+        /// <param name="model"></param>
+        /// <response code="200">Returns created office</response>
```
Comment (kseniya-kalita) [inline OfficeMicroService/Presentation/Controllers/OfficeController.cs]: not really created:)

File: OfficeMicroService/Presentation/Controllers/OfficeController.cs [hunk 24]
```diff
@@ -0,0 +1,129 @@
...
+        public async Task<IActionResult> Update(string id, OfficeDTO model)
+        {
+            Log.Information("Method Update {0}", id);
+            var office = await _officeServices.UpdateAsync(id, model);
+            if (office == null)
+            {
+                Log.Information("Method Update NotFound {0}", id);
+                return NotFound();
+            }
+            return Ok(office);
```
Comment (kseniya-kalita) [inline OfficeMicroService/Presentation/Controllers/OfficeController.cs]: what other status code can be returned in Put?

File: OfficeMicroService/Presentation/Controllers/OfficeController.cs [hunk 25]
```diff
@@ -0,0 +1,129 @@
...
+        /// </summary>
+        /// <param name="id"></param>
+        /// <response code="200">Returns updated office</response>
+        /// <response code="401">Unauthorized</response>
+        /// <response code="403">Forbidden</response>
+        /// <response code="404">Office not found</response>
+        /// <response code="500">Operation wasn't succeeded</response>
+        [Authorize(Roles = nameof(UserRole.Receptionist))]
+        [HttpPut("{id}")]
+        public async Task<IActionResult> MakeInactive(string id)
```
Comment (kseniya-kalita) [inline OfficeMicroService/Presentation/Controllers/OfficeController.cs]: try to merge with the methd MakeActive not to duplicate the logic

File: OfficeMicroService/appsettings.json [hunk 26]
```diff
@@ -0,0 +1,24 @@
...
+        "Name": "File",
+        "Args": { "path": "Logs/log.txt" }
+      }
+    ]
+  },
+  "AllowedHosts": "*",
+  "OfficeStoreDatabaseSettings": {
+    "OfficesCollectionName": "Office",
+    "ConnectionString": "mongodb+srv://Kikusha:31581520@officemicrodb.39sirld.mongodb.net/?retryWrites=true&w=majority",
+    "DatabaseName": "OfficeMIcroDB"
```
Comment (kseniya-kalita) [inline OfficeMicroService/appsettings.json]: probaby I should be in lower case

File: OfficeMicroService/Program.cs [hunk 27]
```diff
@@ -0,0 +1,39 @@
...
+
+namespace OfficeMicroService
+{
+    public class Program
+    {
+        public static void Main(string[] args)
+        {
+            var builder = WebApplication.CreateBuilder(args);
+
+            builder.Services.ConfigureJWTAuthentication(builder.Configuration);
```
Comment (kseniya-kalita) [inline OfficeMicroService/Program.cs]: extract builder.Configuration to variable

**Review by kseniya-kalita — CHANGES_REQUESTED** [review]
Looks good, need to resolve mostly cleanup and small refactoring changes. Main points to resolve are enpoints for MakeActive/MakeInactive and handle audience validation

File: OfficeMicroService/Application/Services/OfficeServices.cs (hunk 8 above)
Comment (NikDub) [reply OfficeMicroService/Application/Services/OfficeServices.cs]: FindAsync() generate exception when we have not entity with id

File: OfficeMicroService/Presentation/Controllers/OfficeController.cs (hunk 24 above)
Comment (NikDub) [reply OfficeMicroService/Presentation/Controllers/OfficeController.cs]: 201 or 204

File: OfficeMicroService/Application/Services/OfficeServices.cs (hunk 8 above)
Comment (kseniya-kalita) [reply OfficeMicroService/Application/Services/OfficeServices.cs]: can you then catch not generic Exception but check that this is exactly that exception?

File: OfficeMicroService/Application/Exceptions/UnauthorizedException.cs [hunk 28]
```diff
@@ -0,0 +1,9 @@
+﻿namespace OfficeMicroService.Application.Exceptions
+{
+    public class UnauthorizedException : Exception
+    {
+        public UnauthorizedException() : base("Unauthorized user.")
```
Comment (kseniya-kalita) [inline OfficeMicroService/Application/Exceptions/UnauthorizedException.cs]: Do you use it?

File: OfficeMicroService/Data/Models/DTO/OfficeDTO.cs (hunk 13 above)
Comment (NikDub) [reply OfficeMicroService/Data/Models/DTO/OfficeDTO.cs]: only Required?

File: OfficeMicroService/Presentation/Controllers/OfficeController.cs (hunk 18 above)
Comment (NikDub) [reply OfficeMicroService/Presentation/Controllers/OfficeController.cs:28]: i dont understand...

File: OfficeMicroService/Presentation/Controllers/OfficeController.cs (hunk 20 above)
Comment (NikDub) [reply OfficeMicroService/Presentation/Controllers/OfficeController.cs]: photo save in another microservices,  in this we have only photoId

File: OfficeMicroService/Application/Extensions/ServiceExtensions.cs [hunk 29]
```diff
@@ -0,0 +1,85 @@
...
+namespace OfficeMicroService.Application.Extensions
+{
+    public static class ServiceExtensions
+    {
+        public static void ConfigureJWTAuthentication(this IServiceCollection services, IConfiguration configuration)
+        {
+            services.AddAuthentication(JwtBearerDefaults.AuthenticationScheme)
+                   .AddJwtBearer(JwtBearerDefaults.AuthenticationScheme, options =>
+                   {
+                       options.Authority = configuration.GetValue<string>("Routes:AuthorityRoute");
```
Comment (kseniya-kalita) [inline OfficeMicroService/Application/Extensions/ServiceExtensions.cs]: much better, but what if configuration.GetValue will return null?

File: OfficeMicroService/Application/Extensions/ServiceExtensions.cs [hunk 30]
```diff
@@ -0,0 +1,85 @@
...
+                       options.Audience = configuration.GetValue<string>("Routes:Scopes");
+                       options.TokenValidationParameters = new TokenValidationParameters
+                       {
+                           ValidateAudience = false,
```
Comment (kseniya-kalita) [inline OfficeMicroService/Application/Extensions/ServiceExtensions.cs]: I don't think that not validating Audience is secure

File: OfficeMicroService/Application/Extensions/ServiceExtensions.cs [hunk 31]
```diff
@@ -0,0 +1,85 @@
...
+        public static void ConfigureDbConnection(this IServiceCollection services, IConfiguration configuration)
+        {
+            services.Configure<OfficeStoreDatabaseSettings>(configuration.GetSection(nameof(OfficeStoreDatabaseSettings)));
+
+            services.AddSingleton<IOfficeStoreDatabaseSettings>(sp =>
+                sp.GetRequiredService<IOptions<OfficeStoreDatabaseSettings>>().Value);
+        }
+        public static void ConfigureServices(this IServiceCollection services)
+        {
+            services.AddSingleton<IOfficeServices, OfficeServices>();
```
Comment (kseniya-kalita) [inline OfficeMicroService/Application/Extensions/ServiceExtensions.cs]: Usually services with bussiness logic and repos are Scoped. Investigate the difference between AddScoped, AddTransient and AddSingleton and choose the right here

File: OfficeMicroService/Presentation/Controllers/OfficeController.cs [hunk 32]
```diff
@@ -0,0 +1,129 @@
...
+        /// <response code="404">Office not found</response>
+        /// <response code="500">Operation wasn't succeeded</response>
+        [HttpGet("{id}")]
+        public async Task<IActionResult> GetById(string id)
+        {
+            Log.Information("Method {0} {1}", nameof(GetById), id);
+            var office = await _officeServices.GetAsync(id);
+            if (office == null)
+            {
+                Log.Information("Method GetById NotFound {0}", id);
```
Comment (kseniya-kalita) [inline OfficeMicroService/Presentation/Controllers/OfficeController.cs]: It is not really thay the method not found but office with provided id not found. Please change this and Add the same message for BadRequest

File: OfficeMicroService/Presentation/Controllers/OfficeController.cs [hunk 33]
```diff
@@ -0,0 +1,129 @@
...
+            {
+                return BadRequest(model);
+            }
+            var office = await _officeServices.UpdateAsync(id, model);
+            if (office == null)
+            {
+                Log.Information("Method {0} NotFound {1}", nameof(Update), id);
+                return NotFound();
+            }
+            return Created("", office);
```
Comment (kseniya-kalita) [inline OfficeMicroService/Presentation/Controllers/OfficeController.cs]: Created is not suitable for Put

File: OfficeMicroService/Presentation/Controllers/OfficeController.cs [hunk 34]
```diff
@@ -0,0 +1,129 @@
...
+        /// <response code="403">Forbidden</response>
+        /// <response code="404">Office not found</response>
+        /// <response code="500">Operation wasn't succeeded</response>
+        [Authorize(Roles = nameof(UserRole.Receptionist))]
+        [HttpPut("{id}/status")]
+        public async Task<IActionResult> ChangeStatus(string id)
+        {
+            Log.Information("Method {0} {1}", nameof(ChangeStatus), id);
+            await _officeServices.ChangeStatus(id);
+            return Created("", null);
```
Comment (kseniya-kalita) [inline OfficeMicroService/Presentation/Controllers/OfficeController.cs]: Created is not suitable for Put

File: OfficeMicroService/Presentation/Controllers/OfficeController.cs [hunk 35]
```diff
@@ -0,0 +1,129 @@
...
+
+        /// <summary>
+        /// Update office status for Receptionist
+        /// </summary>
+        /// <param name="id"></param>
+        /// <param name="status"></param>
+        /// <response code="200">Returns updated office</response>
+        /// <response code="401">Unauthorized</response>
+        /// <response code="403">Forbidden</response>
+        /// <response code="404">Office not found</response>
```
Comment (kseniya-kalita) [inline OfficeMicroService/Presentation/Controllers/OfficeController.cs:121]: You should change the services or controller implementation to return 404

File: OfficeMicroService/Application/Services/OfficeServices.cs [hunk 36]
```diff
@@ -0,0 +1,91 @@
...
+        {
+
+            var office = await GetAsync(id);
+            if (office == null)
+                return null;
+
+            OfficeStatus status;
+            if (Enum.TryParse(office.Status, out status))
+            {
+                status++;
```
Comment (kseniya-kalita) [inline OfficeMicroService/Application/Services/OfficeServices.cs]: It's not really understandable from Bussiness logic prospective so please change it

File: OfficeMicroService/Application/Services/OfficeServices.cs [hunk 37]
```diff
@@ -0,0 +1,91 @@
...
+                return null;
+            }
+        }
+
+        public async Task RemoveAsync(string id)
+        {
+            await _offices.DeleteOneAsync(office => office.Id == id);
+        }
+
+        public async Task<Office> ChangeStatus(string id)
```
Comment (kseniya-kalita) [inline OfficeMicroService/Application/Services/OfficeServices.cs]: Async ending

File: OfficeMicroService/Application/Services/OfficeServices.cs [hunk 38]
```diff
@@ -0,0 +1,91 @@
...
+            var database = client.GetDatabase(settings.DatabaseName);
+
+            _offices = database.GetCollection<Office>(settings.OfficesCollectionName);
+            _mapper = mapper;
+        }
+
+        public async Task<List<OfficeList>> GetAllAsync()
+        {
+            var officeList = (await _offices.FindAsync(office => true)).ToList();
+            var mapModel = _mapper.Map<List<OfficeList>>(officeList);
```
Comment (kseniya-kalita) [inline OfficeMicroService/Application/Services/OfficeServices.cs]: Can you please investigate if a Map method could return null? Not sure about that. It's better to check officeList for null, not mapModel

File: OfficeMicroService/Application/Services/OfficeServices.cs [hunk 39]
```diff
@@ -0,0 +1,91 @@
...
+            if (mapModel == null)
```
Comment (kseniya-kalita) [inline OfficeMicroService/Application/Services/OfficeServices.cs]: It looks like that this code is redudant

File: OfficeMicroService/Data/Models/Office.cs [hunk 40]
```diff
@@ -0,0 +1,14 @@
+﻿using MongoDB.Bson;
+using MongoDB.Bson.Serialization.Attributes;
+using OfficeMicroService.Data.Models.DTO;
+
+namespace OfficeMicroService.Data.Models
+{
+    public class Office : OfficeDTO
```
Comment (kseniya-kalita) [inline OfficeMicroService/Data/Models/Office.cs]: Nope, remove inheritance

File: OfficeMicroService/Data/Models/OfficeList.cs [hunk 41]
```diff
@@ -0,0 +1,13 @@
+﻿namespace OfficeMicroService.Data.Models
+{
+    public class OfficeList
```
Comment (kseniya-kalita) [inline OfficeMicroService/Data/Models/OfficeList.cs]: The naming is confusing as there is no list inside. Please find better naming

File: OfficeMicroService/Data/Models/DTO/OfficeDTO.cs [hunk 42]
```diff
@@ -0,0 +1,25 @@
+﻿using MongoDB.Bson.Serialization.Attributes;
+using MongoDB.Bson;
+using System.ComponentModel.DataAnnotations;
+
+namespace OfficeMicroService.Data.Models.DTO
+{
+    public class OfficeDTO
```
Comment (kseniya-kalita) [inline OfficeMicroService/Data/Models/DTO/OfficeDTO.cs]: Please find better place for DTOs. Core (Models) is for Entities, but not DTOs

File: OfficeMicroService/Application/Services/OfficeServices.cs [hunk 43]
```diff
@@ -0,0 +1,91 @@
...
+                return null;
+
+            return mapModel;
+        }
+
+        public async Task<Office> GetAsync(string id)
```
Comment (kseniya-kalita) [inline OfficeMicroService/Application/Services/OfficeServices.cs]: Services should accept and return DTOs, entities are hidden under the repo level

File: OfficeMicroService/Application/Services/OfficeServices.cs [hunk 44]
```diff
@@ -0,0 +1,80 @@
...
+        {
+            var mapModel = _mapper.Map<OfficeIdDTO>(model);
+            mapModel.Id = id;
+
+            try
+            {
+                await _offices.ReplaceOneAsync(Office => Office.Id == id, mapModel);
+                return mapModel;
+            }
+            catch (Exception)
```
Comment (kseniya-kalita) [inline OfficeMicroService/Application/Services/OfficeServices.cs]: Same thing with Exception as above.

File: OfficeMicroService/Application/Services/OfficeServices.cs [hunk 45]
```diff
@@ -0,0 +1,91 @@
...
+using OfficeMicroService.Data.Models.DTO;
+using OfficeMicroService.Data.Settings;
+using System.Data.Common;
+using System.Runtime.InteropServices;
+
+namespace OfficeMicroService.Application.Services
+{
+    public class OfficeServices : IOfficeServices
+    {
+        private readonly IMongoCollection<Office> _offices;
```
Comment (kseniya-kalita) [inline OfficeMicroService/Application/Services/OfficeServices.cs]: Let's move the logic the work with with IMongoCollection to repo, it shouldn't take a lot of time but the code will be less coupled

**Review by kseniya-kalita — CHANGES_REQUESTED** [review]
Please get acquinted with changes and resolve comments. Good work on resolving previous conversations! Main point is to add extra repositories layer so we can decouple our service layer with MongoDb Collection

File: OfficeMicroService/Application/Exceptions/UnauthorizedException.cs (hunk 28 above)
Comment (NikDub) [reply OfficeMicroService/Application/Exceptions/UnauthorizedException.cs]: yes

File: OfficeMicroService/Application/Services/OfficeServices.cs [hunk 46]
```diff
@@ -0,0 +1,83 @@
...
+        {
+
+            var office = await GetAsync(id);
+            if (office == null)
+                throw new NotFoundException("Office not found.");
+
+            OfficeStatus status;
+            if (Enum.TryParse(office.Status, out status))
+            {
+                if (status == OfficeStatus.Active)
```
Comment (kseniya-kalita) [inline OfficeMicroService/Application/Services/OfficeServices.cs]: let's do it in one line

File: OfficeMicroService/Application/Services/OfficeServices.cs [hunk 47]
```diff
@@ -0,0 +1,83 @@
...
+                    office.Status = OfficeStatus.Inactive.ToString();
+                else
+                    office.Status = OfficeStatus.Active.ToString();
+
+                office = await UpdateAsync(id, _mapper.Map<OfficeForUpdateDTO>(office));
```
Comment (kseniya-kalita) [inline OfficeMicroService/Application/Services/OfficeServices.cs]: you are calling GetAsync method twice here. Let's better call UpdateAsync method from repo here

File: OfficeMicroService/Presentation/Controllers/OfficeController.cs [hunk 48]
```diff
@@ -0,0 +1,131 @@
...
+        /// <param name="model"></param>
+        /// <response code="204">Returns if office was updated</response>
+        /// <response code="422">Incorrect model</response>
+        /// <response code="401">Unauthorized</response>
+        /// <response code="403">Forbidden</response>
+        /// <response code="404">Office not found</response>
+        /// <response code="500">Operation wasn't succeeded</response>
+        [Authorize(Roles = nameof(UserRole.Receptionist))]
+        [HttpPut("{id}")]
+        public async Task<IActionResult> Update(string id, OfficeForUpdateDTO model)
```
Comment (kseniya-kalita) [inline OfficeMicroService/Presentation/Controllers/OfficeController.cs]: Let's add [FromBody] attribute for better readability

File: OfficeMicroService/Presentation/Controllers/OfficeController.cs [hunk 49]
```diff
@@ -0,0 +1,131 @@
...
+            }
+            return Ok(office);
+        }
+
+        /// <summary>
+        /// Create office for Receptionist
+        /// </summary>
+        /// <param name="model"></param>
+        /// <response code="201">Returns created office</response>
+        /// <response code="422">Incorrect model</response>
```
Comment (kseniya-kalita) [inline OfficeMicroService/Presentation/Controllers/OfficeController.cs]: Let's move this line down to have status codes written incrementally (201, 401, 403, 422, 500)

File: OfficeMicroService/Application/Services/DTO/OfficeForChangeDTO.cs [hunk 50]
```diff
@@ -0,0 +1,27 @@
+﻿using MongoDB.Bson.Serialization.Attributes;
+using MongoDB.Bson;
+using System.ComponentModel.DataAnnotations;
+
+namespace OfficeMicroService.Application.Services.DTO
```
Comment (kseniya-kalita) [inline OfficeMicroService/Application/Services/DTO/OfficeForChangeDTO.cs]: Good job! But let's move DTOs folder upper to the Application folder, not inside the Services folder

File: OfficeMicroService/Application/Services/DTO/OfficeForChangeDTO.cs [hunk 51]
```diff
@@ -0,0 +1,27 @@
...
+{
+    public class OfficeForChangeDTO
+    {
+        private const string NumberRegex = "[+]{1}[0-9]{12}";
+
+        [BsonRepresentation(BsonType.ObjectId)]
+        public string PhotoId { get; set; }
+        [BsonRequired]
```
Comment (kseniya-kalita) [inline OfficeMicroService/Application/Services/DTO/OfficeForChangeDTO.cs]: It's not possible to check this with ModelState required so I was able to create an entity with all field equal to null, please fix. Entity is is 63a084b7dd56614723d3c51a

File: OfficeMicroService/Application/Services/OfficeServices.cs [hunk 52]
```diff
@@ -0,0 +1,83 @@
...
+        }
+        public async Task<OfficeDTO> GetAsync(string id)
+        {
+            var office = (await _repository.FindByCondition(office => office.Id == id)).FirstOrDefault();
+            return _mapper.Map<OfficeDTO>(office);
+        }
+        public async Task<OfficeDTO> CreateAsync(OfficeForChangeDTO model)
+        {
+            var mapModel = _mapper.Map<Office>(model);
+            if (mapModel == null)
```
Comment (kseniya-kalita) [inline OfficeMicroService/Application/Services/OfficeServices.cs]: The only situation when it's null is when model is null. So you need to remove this map & check and add check if model is null (before mapping) . Please change it in other places also

**Review by kseniya-kalita — COMMENTED** [review]
Good job, looks much better, just few points to resolve

**Pull request merged**